planner/core/name_resolution.go

```diff
@@ -0,0 +1,884 @@
+	// down in corresponding scope which will be used in later in formal plan building phase, along with correlated column within them.
+	//
+	p LogicalPlan
+}
+
+// ScopeSchema is used to resolve column and register basic agg in analyzing phase.
+type ScopeSchema struct {
+	scopeSchema *expression.Schema
+	scopeNames  []*types.FieldName
+
+	// agg group utility elements, aggFuncs are mapped to aggColumn and aggFuncExpr, and aggMapper is used to fast locate the offset in
+	// aggFuncs/aggColumn when an address of *AggregateFuncExpr is given.
+	aggFuncs     []*aggregation.AggFuncDesc
+	aggColumn    []*expression.Column
+	astAggFunc   []*ast.AggregateFuncExpr
+	aggMapper    map[*ast.AggregateFuncExpr]int
+	groupByItems []expression.Expression
+
+	// win group utility elements, windowFuncs are mapped to windowColumn, and windowMapper is used to fast locate the
+	// offset in windowFuncs/windowColumn when an address of *WindowFuncExpr is given.
+	// todo：change window function resolution as the newer one. 2022/06/30.
+	// windowFuncs  []*windowFuncs
+	// windowColumn []*expression.Column
+	// windowMapper map[*ast.WindowFuncExpr]int
+
+	// we should build the projection expr out and map them to a specific column in analyzing phase.
+	// otherwise, cases like: select (select 1) as a from dual order by a & select 1 as a, (select t.a) from t
+	// they can't resolve themselves (mainly a) from the base from scope columns here.
+	projExpr   []expression.Expression
+	projColumn []*expression.Column
+	projNames  []*types.FieldName
+
+	mapScalarSubQueryByAddr map[*ast.SubqueryExpr]*preBuiltSubQueryCacheItem
+
+	// since tidb's plan building is from bottom up，every operator only output what they care about.
+	// generally speaking, current building process is from data source -> selection -> agg -> projection -> distinct -> order by...
+	// actually, data source & selection & agg won't impose any elimination on the source schema, which means the predicate on
+	// them can refer to any columns that derived from the base table. While after projection, the schema is projected, which causing
+	// the predicate in order by can't refer to some base col if you don't notify projection operator to keep it for you in advance.
+	//
+	// that's why we analyze these clauses first before we build it, and using this fields to tell projection to reserve it for later use.
+	reservedCols                []*expression.Column
+	reservedCorrelatedCols      []*expression.CorrelatedColumn
+	projectionCol4CorrelatedAgg map[int64]*expression.Column
+
+	// inAgg is flag that used to judge whether current context is in agg.
+	inAgg bool
+}
+
+// GetCol is used to get column by unique id.
+func (s *ScopeSchema) GetCol(id int64) (*expression.Column, int) {
+	for i, one := range s.scopeSchema.Columns {
+		if one.UniqueID == id {
+			return one, i
+		}
+	}
+	return nil, -1
```

Comment:
In the future, we can implement the `FastIntMap` based on the `FastIntSet` to speed up this func. It would be useful for wide table.